event/src/main/java/com/bazaarvoice/emodb/event/db/astyanax/AstyanaxEventWriterDAO.java

```diff
@@ -79,12 +90,16 @@ public Void apply(BatchUpdate.Row<ByteBuffer, Integer> row) {
                             });
 
                     for (int i = 0; i < allocation.getLength(); i++) {
-                        int eventIdx = allocation.getOffset() + i;
-                        ByteBuffer eventData = eventsIter.next();
-                        row.putColumn(eventIdx, eventData, Ttls.toSeconds(eventTtl, 1, null));
-
-                        if (sink != null && !sink.accept(AstyanaxEventId.create(channel, slabId, eventIdx), eventData)) {
-                            sink = null;  // Sink isn't interested in more events.
+                        if (eventsIter.hasNext()) {
```

Comment:
Well we could throw an exception here, or just not check it at all and let an exception get thrown on a null pointer etc. It just seems a bad idea to me to not check if the iterator has events here, and if we exit when the iterator is empty I don't think any harm is done. Now I agree there should never be a discrepancy, but this code is doing weird things. It is doing an outer for loop to get a one up counter and then advancing an iterator inside that loop. Maybe the thing to do would be to set a variable one-up counter value (eventIdx) to zero then have a while loop advance the iterator and upon each successful advance increment the one-up counter inside the while loop. Then there will be no need to verify the iterator as the loop will be driven by the iterator state (hasNext) and the one-up counter value will be computed rather than being a loop variable.

Reply:
My feeling is that I would always rather err on the side of failure (I.e. an NPE being thrown) rather than risk silent corruption. @sujithvaddi @mike-unitskyi What are your thoughts on this?

Reply:
Oh I forgot about this. I am switching this to be like I said above where the outer loop is done on the event iterator and eventIdx is computed then there is no weird thing going on where we are advancing an iterator based on a for loop.

Reply:
I am also going to add an explicit check to verify that the iterator had as many elements as the allocation and throw a RuntimeException if it doesn't. I like that better than throwing an NPE, those are the worst because it can be so unclear what caused them. This will be explicit so some poor future maintainer (perhaps me in a few months!) will know what caused the exception.